src/main/java/com/hms/flexyosisoftconnector/OSIsoftConnectorMain.java

```diff
@@ -170,78 +179,105 @@ private static void getNewPayloadToSend() {
+        if (queuePoints.size() != 0) {
+          // Check if queue is behind
+          try {
+            long queueBehindMillis = HistoricalDataQueueManager.getQueueTimeBehindMillis();
+            if (queueBehindMillis >= OSIsoftConfig.WARNING_LIMIT_QUEUE_BEHIND_MILLISECONDS) {
+              Logger.LOG_WARN(
+                  "The historical data queue is running behind by "
+                      + SCTimeUtils.getDayHourMinSecsForMillis(queueBehindMillis));
+            }
+          } catch (IOException e) {
+            Logger.LOG_SERIOUS("Unable to detect if historical data queue is running behind.");
+            Logger.LOG_EXCEPTION(e);
+          } catch (TimeTrackerUnrecoverableException e) {
+            Logger.LOG_SERIOUS("Time tracker file has been corrupted.");
+            Logger.LOG_EXCEPTION(e);
+          }
 
-          // build a payload out of data points
-          if (data != null) {
-            payloadMngr.appendDataPointToPayLoad(data);
+          Logger.LOG_DEBUG("Grabbed " + queuePoints.size() + " datapoints for proccessing.");
+          for (int i = 0; i < queuePoints.size(); i++) {
+            DataPoint data = ((DataPoint) queuePoints.get(i));
+
+            // build a payload out of data points
+            if (data != null) {
+              payloadMngr.appendDataPointToPayLoad(data);
+            }
+
+            // Sleep between datapoints to significantly increase Flexy performance.
+            Thread.yield();
+            try {
+              final int threadWaitMS = 5;
+              Thread.sleep(threadWaitMS);
+            } catch (InterruptedException e) {
+              Logger.LOG_EXCEPTION(e);
+              Logger.LOG_SERIOUS("Exception thrown while sleeping thread.");
+            }
           }
+        }
+      } else {
+        payloadQueueFullCount += 1;
 
-          // Sleep between datapoints to significantly increase Flexy performance.
-          Thread.yield();
-          try {
-            final int threadWaitMS = 5;
-            Thread.sleep(threadWaitMS);
-          } catch (InterruptedException e) {
-            Logger.LOG_EXCEPTION(e);
-            Logger.LOG_SERIOUS("Exception thrown while sleeping thread.");
+        if (payloadQueueFullCount >= payloadQueueFullWarnThresh) {
+          payloadQueueFullCount = 0;
+          Logger.LOG_DEBUG("Reading historical data skipped due to full payload queue.");
+
+          // after warning increase the threshold, but avoid roll over
+          if (payloadQueueFullWarnThresh < Long.MAX_VALUE / 100) {
+            payloadQueueFullWarnThresh *= 10;
           }
         }
+        // This thread should yield and give time for publisher thread to decrease queue size
+        Thread.yield();
+        try {
+          Thread.sleep(FULL_QUEUE_DELAY_MS);
+        } catch (InterruptedException e) {
+          Logger.LOG_EXCEPTION(e);
+          Logger.LOG_SERIOUS("Exception thrown while sleeping thread.");
+        }

```

Comment:
What makes this necessary, and was it tested without? 

I am hesitant to introduce an additional Thread.sleep() here. This code should be executing on the main thread, which means we can/should do nothing and fall back to the standard delay managed by the main method.

Reply:
See

https://github.com/hms-networks/flexy-osisoft-connector/blob/dev/queue/src/main/java/com/hms/flexyosisoftconnector/OSIsoftConnectorMain.java#L139-L150

Reply:
I have not addressed this comment